### PR TITLE
[MklEinsumop] reorder the weights into mkl primitive requested format

### PR DESCRIPTION
**Pull request by snadampal**

### CODE REVIEW

**Comment by snadampal** [comment]
I see that ARM CI also passed with this commit:
https://github.com/tensorflow/tensorflow/actions/runs/3860705344

**Comment by elfringham** [comment]
My local testing shows that the failing tests will pass with this change.

**Comment by penpornk** [comment]
Thank you for the fix! Adding context just for future reference.
This PR fixes 7 failing unit tests introduced by PR https://github.com/tensorflow/tensorflow/pull/59097
```
//bazel_pip/tensorflow/python/ops/parallel_for:math_test_cpu
//bazel_pip/tensorflow/python/ops/parallel_for:math_test_gpu
//bazel_pip/tensorflow/python/kernel_tests/linalg:einsum_op_test_cpu
//bazel_pip/tensorflow/python/kernel_tests/linalg:einsum_op_test_gpu
//bazel_pip/tensorflow/python:special_math_ops_test_cpu
//bazel_pip/tensorflow/python:special_math_ops_test_gpu
//bazel_pip/tensorflow/dtensor/python/tests:spmd_test_cpu
```

### Example test outputs
Test output for `//bazel_pip/tensorflow/python/kernel_tests/linalg:einsum_op_test_cpu`
```
2023-01-04T17:58:41.2807608Z 2023-01-04 17:58:14.677100: I tensorflow/core/common_runtime/executor.cc:1195] [/job:localhost/replica:0/task:0/device:CPU:0] Executor start aborting: INVALID_ARGUMENT: Expected input 0 to have rank 2 but got: 3
2023-01-04T17:58:41.2809186Z 	 [[{{node Einsum_1}}]]
2023-01-04T17:58:41.2810265Z ERROR:tensorflow:Graph execution error:
...
2023-01-04T17:58:41.3754884Z 2023-01-04 17:58:15.055922: I tensorflow/core/common_runtime/executor.cc:1195] [/job:localhost/replica:0/task:0/device:CPU:0] Executor start aborting: INVALID_ARGUMENT: Expected input 0 to have rank at least 2 but got: 1
2023-01-04T17:58:41.3755555Z 	 [[{{node Einsum_3}}]]
2023-01-04T17:58:41.3755923Z ERROR:tensorflow:Graph execution error:
...
2023-01-04T17:58:41.3888053Z 2023-01-04 17:58:15.069017: W tensorflow/core/framework/op_kernel.cc:1830] OP_REQUIRES failed at einsum_op_impl.h:498 : INVALID_ARGUMENT: Expected dimension 3 at axis 0 of the input shaped [4,4] but got dimension 4
2023-01-04T17:58:41.3889238Z 2023-01-04 17:58:15.069057: I tensorflow/core/common_runtime/executor.cc:1195] [/job:localhost/replica:0/task:0/device:CPU:0] Executor start aborting: INVALID_ARGUMENT: Expected dimension 3 at axis 0 of the input shaped [4,4] but got dimension 4
2023-01-04T17:58:41.3889903Z 	 [[{{node Einsum_5}}]]
```

Test output for `//bazel_pip/tensorflow/python/ops/parallel_for:math_test_cpu`
```
2023-01-04T17:58:10.0376405Z AssertionError: 
2023-01-04T17:58:10.0376855Z Not equal to tolerance rtol=0.0001, atol=1e-05
2023-01-04T17:58:10.0377259Z Mismatched value: a is different from b. 
2023-01-04T17:58:10.0377694Z not close where = (array([0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2,
2023-01-04T17:58:10.0378097Z        2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 4, 4,
2023-01-04T17:58:10.0378464Z        4, 5, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 7, 7, 7,
2023-01-04T17:58:10.0378831Z        7, 7, 7, 7, 7, 7, 8, 8, 8, 8, 8, 8, 8, 8, 8, 9, 9, 9, 9, 9, 9, 9,
2023-01-04T17:58:10.0379218Z        9, 9]), array([0, 1, 2, 3, 4, 5, 6, 7, 8, 0, 1, 2, 3, 4, 5, 6, 7, 8, 0, 1, 2, 3,
2023-01-04T17:58:10.0379598Z        4, 5, 6, 7, 8, 0, 1, 2, 3, 4, 5, 6, 7, 8, 0, 1, 2, 3, 4, 5, 6, 7,
2023-01-04T17:58:10.0379952Z        8, 0, 1, 2, 3, 4, 5, 6, 7, 8, 0, 1, 2, 3, 4, 5, 6, 7, 8, 0, 1, 2,
2023-01-04T17:58:10.0380318Z        3, 4, 5, 6, 7, 8, 0, 1, 2, 3, 4, 5, 6, 7, 8, 0, 1, 2, 3, 4, 5, 6,
2023-01-04T17:58:10.0380700Z        7, 8]), array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
2023-01-04T17:58:10.0381077Z        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
2023-01-04T17:58:10.0381442Z        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
2023-01-04T17:58:10.0381807Z        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
2023-01-04T17:58:10.0382134Z        0, 0]))
2023-01-04T17:58:10.0382595Z not close lhs = [ 6.9418353e-01  5.1605169e-02  1.0278878e+00  3.2460448e-01
2023-01-04T17:58:10.0383087Z   8.8283783e-01  2.8944468e-01  3.4586275e-01  6.8949014e-01
2023-01-04T17:58:10.0383546Z   4.3784395e-01 -4.2130753e+28 -1.2520518e+29 -1.1946045e+29
2023-01-04T17:58:10.0383996Z  -8.7718373e+28 -2.7290371e+29 -2.7367525e+29 -7.7126351e+28
2023-01-04T17:58:10.0384456Z  -2.6753542e+29 -1.6750995e+29  8.7638569e-01  9.3259744e-02
2023-01-04T17:58:10.0384918Z   2.1901062e-01  2.1247660e-01  2.3610549e-01  7.5771093e-01
2023-01-04T17:58:10.0385368Z   9.2133188e-01  4.1847992e-01  4.6887341e-01  9.2336375e-01
2023-01-04T17:58:10.0385823Z   4.0780917e-01  6.8191242e-01  5.7174760e-01  5.5216324e-01
2023-01-04T17:58:10.0386274Z   7.3276585e-01  3.3736566e-01  6.5520406e-01  8.8787621e-01
2023-01-04T17:58:10.0386726Z   8.3954525e-01  9.6947408e-01  3.0326846e-01  4.0657654e-01
2023-01-04T17:58:10.0387170Z   4.0534210e-01  3.2056653e-01  4.1624713e-01  3.8456905e-01
2023-01-04T17:58:10.0387629Z   3.9938489e-01 -1.3876914e+17 -5.3489806e+16 -2.6843281e+17
2023-01-04T17:58:10.0388085Z  -6.9373480e+16 -2.0027970e+17 -1.5428741e+17 -1.9495913e+17
2023-01-04T17:58:10.0388532Z  -6.5686770e+16 -1.7873381e+17  6.8086565e-01  2.3144978e-01
2023-01-04T17:58:10.0388986Z   5.2360278e-01  5.0153357e-01  6.2628537e-01  7.2663552e-01
2023-01-04T17:58:10.0389442Z   4.7623450e-01  8.0457222e-01  1.7693582e-01 -4.0658934e+09
2023-01-04T17:58:10.0390084Z  -5.4855711e+09 -2.8799826e+09 -8.2351058e+09 -6.9627602e+09
2023-01-04T17:58:10.0390601Z  -3.5069665e+09 -7.1533517e+09 -4.0856525e+09 -2.8997317e+09
2023-01-04T17:58:10.0390972Z   2.3063752e+19  1.7716752e+19  2.3656734e+19  1.9443036e+19
2023-01-04T17:58:10.0391335Z   2.2024021e+19  1.4372444e+19  1.0249816e+19  1.2379857e+19
2023-01-04T17:58:10.0391689Z   1.5362732e+19  7.0019644e+24  1.0616546e+25  6.7674740e+24
2023-01-04T17:58:10.0392050Z   6.3910999e+24  5.4547201e+24  1.1781101e+25  4.6182427e+24
2023-01-04T17:58:10.0392587Z   7.2423583e+24  5.3846403e+24]
2023-01-04T17:58:10.0393095Z not close rhs = [-2.35491515e+36 -3.05349656e+35 -2.02785971e+36 -7.10510397e+35
2023-01-04T17:58:10.0393602Z  -2.01079936e+36 -2.08686235e+36 -1.13472684e+35 -2.39250506e+34
2023-01-04T17:58:10.0394074Z  -1.68776918e+36 -6.47776706e+35 -5.11355238e+35 -6.90329083e+35
2023-01-04T17:58:10.0394531Z  -1.25107829e+36 -2.25387231e+36 -2.07686487e+36 -7.77904804e+35
2023-01-04T17:58:10.0395006Z  -2.37355073e+36 -1.31634003e+36 -2.58282352e+36 -9.18606493e+35
2023-01-04T17:58:10.0395471Z  -1.23576760e+36 -2.67270611e+34 -1.09347834e+36 -7.22516554e+35
2023-01-04T17:58:10.0395938Z  -3.87860083e+35 -1.44078584e+36 -1.94091774e+36 -2.31769313e+36
2023-01-04T17:58:10.0396392Z  -1.90787326e+36 -1.80641272e+36 -1.91509799e+35 -8.17256620e+34
2023-01-04T17:58:10.0396854Z  -1.42686324e+36 -1.61725342e+36 -1.51752355e+36 -5.92245410e+35
2023-01-04T17:58:10.0397319Z  -6.86439218e+35 -2.50680901e+36 -1.67371664e+36 -2.26370262e+36
2023-01-04T17:58:10.0397775Z  -6.88866055e+35 -4.20921084e+35 -2.11096562e+36 -2.08122574e+36
2023-01-04T17:58:10.0398245Z  -1.02992594e+36 -2.06762370e+36 -3.68009784e+35 -2.64620969e+36
2023-01-04T17:58:10.0398713Z  -5.08075707e+35 -1.92123208e+36 -6.28624922e+35 -1.11146131e+36
2023-01-04T17:58:10.0399174Z  -1.17867688e+35 -2.43464737e+36 -1.28277106e+36 -4.81338857e+35
2023-01-04T17:58:10.0399638Z  -1.30482128e+36 -4.13324846e+35 -2.52445407e+36 -4.13683790e+35
2023-01-04T17:58:10.0400107Z  -1.24411429e+36 -1.21053304e+36 -7.02579024e+35 -1.13227708e+36
2023-01-04T17:58:10.0400571Z  -1.35343933e+36 -2.45768739e+36 -1.13636169e+36 -1.81353185e+36
2023-01-04T17:58:10.0401028Z  -2.11912111e+34 -1.81213315e+36 -1.97982511e+36 -3.84766976e+35
2023-01-04T17:58:10.0401492Z  -1.03467163e+36 -1.24179112e+35 -2.24944488e+36 -2.16362826e+36
2023-01-04T17:58:10.0401958Z  -5.44790711e+35 -7.00975288e+35 -1.48751699e+36 -9.58796009e+35
2023-01-04T17:58:10.0402415Z  -6.08439173e+33 -6.28868152e+35 -3.73592557e+35 -9.27910177e+35
2023-01-04T17:58:10.0402885Z  -1.54416782e+36 -8.95541987e+35 -2.35344230e+36 -7.81864786e+35
2023-01-04T17:58:10.0403295Z  -2.06935150e+36 -2.15092716e+35]
2023-01-04T17:58:10.0403668Z not close dif = [2.35491515e+36 3.05349656e+35 2.02785971e+36 7.10510397e+35
2023-01-04T17:58:10.0404060Z  2.01079936e+36 2.08686235e+36 1.13472684e+35 2.39250506e+34
2023-01-04T17:58:10.0404423Z  1.68776918e+36 6.47776666e+35 5.11355120e+35 6.90328924e+35
2023-01-04T17:58:10.0404792Z  1.25107821e+36 2.25387199e+36 2.07686455e+36 7.77904724e+35
2023-01-04T17:58:10.0405147Z  2.37355041e+36 1.31633987e+36 2.58282352e+36 9.18606493e+35
2023-01-04T17:58:10.0405508Z  1.23576760e+36 2.67270611e+34 1.09347834e+36 7.22516554e+35
2023-01-04T17:58:10.0405870Z  3.87860083e+35 1.44078584e+36 1.94091774e+36 2.31769313e+36
2023-01-04T17:58:10.0406225Z  1.90787326e+36 1.80641272e+36 1.91509799e+35 8.17256620e+34
2023-01-04T17:58:10.0406583Z  1.42686324e+36 1.61725342e+36 1.51752355e+36 5.92245410e+35
2023-01-04T17:58:10.0406941Z  6.86439218e+35 2.50680901e+36 1.67371664e+36 2.26370262e+36
2023-01-04T17:58:10.0407297Z  6.88866055e+35 4.20921084e+35 2.11096562e+36 2.08122574e+36
2023-01-04T17:58:10.0407654Z  1.02992594e+36 2.06762370e+36 3.68009784e+35 2.64620969e+36
2023-01-04T17:58:10.0408015Z  5.08075707e+35 1.92123208e+36 6.28624922e+35 1.11146131e+36
2023-01-04T17:58:10.0408376Z  1.17867688e+35 2.43464737e+36 1.28277106e+36 4.81338857e+35
2023-01-04T17:58:10.0408725Z  1.30482128e+36 4.13324846e+35 2.52445407e+36 4.13683790e+35
2023-01-04T17:58:10.0409290Z  1.24411429e+36 1.21053304e+36 7.02579024e+35 1.13227708e+36
2023-01-04T17:58:10.0409677Z  1.35343933e+36 2.45768739e+36 1.13636169e+36 1.81353185e+36
2023-01-04T17:58:10.0410034Z  2.11912111e+34 1.81213315e+36 1.97982511e+36 3.84766976e+35
2023-01-04T17:58:10.0410395Z  1.03467163e+36 1.24179112e+35 2.24944488e+36 2.16362826e+36
2023-01-04T17:58:10.0410758Z  5.44790711e+35 7.00975288e+35 1.48751699e+36 9.58796009e+35
2023-01-04T17:58:10.0411111Z  6.08439173e+33 6.28868152e+35 3.73592557e+35 9.27910177e+35
2023-01-04T17:58:10.0411639Z  1.54416782e+36 8.95541987e+35 2.35344230e+36 7.81864786e+35
2023-01-04T17:58:10.0412126Z  2.06935150e+36 2.15092716e+35]
2023-01-04T17:58:10.0412489Z not close tol = [2.35491515e+32 3.05349637e+31 2.02785971e+32 7.10510366e+31
2023-01-04T17:58:10.0412878Z  2.01079935e+32 2.08686226e+32 1.13472679e+31 2.39250500e+30
2023-01-04T17:58:10.0413238Z  1.68776915e+32 6.47776691e+31 5.11355215e+31 6.90329042e+31
2023-01-04T17:58:10.0413605Z  1.25107826e+32 2.25387217e+32 2.07686473e+32 7.77904789e+31
2023-01-04T17:58:10.0413960Z  2.37355060e+32 1.31634004e+32 2.58282340e+32 9.18606443e+31
2023-01-04T17:58:10.0414323Z  1.23576755e+32 2.67270590e+30 1.09347834e+32 7.22516547e+31
2023-01-04T17:58:10.0414829Z  3.87860082e+31 1.44078580e+32 1.94091764e+32 2.31769301e+32
2023-01-04T17:58:10.0415185Z  1.90787315e+32 1.80641274e+32 1.91509796e+31 8.17256582e+30
2023-01-04T17:58:10.0415548Z  1.42686323e+32 1.61725338e+32 1.51752348e+32 5.92245409e+31
2023-01-04T17:58:10.0415908Z  6.86439202e+31 2.50680904e+32 1.67371660e+32 2.26370257e+32
2023-01-04T17:58:10.0416267Z  6.88866049e+31 4.20921084e+31 2.11096553e+32 2.08122576e+32
2023-01-04T17:58:10.0416672Z  1.02992588e+32 2.06762370e+32 3.68009786e+31 2.64620960e+32
2023-01-04T17:58:10.0417033Z  5.08075689e+31 1.92123207e+32 6.28624888e+31 1.11146125e+32
2023-01-04T17:58:10.0417392Z  1.17867680e+31 2.43464739e+32 1.28277107e+32 4.81338844e+31
2023-01-04T17:58:10.0417746Z  1.30482130e+32 4.13324823e+31 2.52445394e+32 4.13683777e+31
2023-01-04T17:58:10.0418112Z  1.24411427e+32 1.21053301e+32 7.02578991e+31 1.13227702e+32
2023-01-04T17:58:10.0418474Z  1.35343926e+32 2.45768739e+32 1.13636164e+32 1.81353187e+32
2023-01-04T17:58:10.0418829Z  2.11912108e+30 1.81213319e+32 1.97982513e+32 3.84766972e+31
2023-01-04T17:58:10.0419194Z  1.03467164e+32 1.24179107e+31 2.24944479e+32 2.16362827e+32
2023-01-04T17:58:10.0419556Z  5.44790718e+31 7.00975278e+31 1.48751697e+32 9.58795973e+31
2023-01-04T17:58:10.0419922Z  6.08439142e+29 6.28868124e+31 3.73592557e+31 9.27910142e+31
2023-01-04T17:58:10.0420282Z  1.54416782e+32 8.95541976e+31 2.35344220e+32 7.81864747e+31
2023-01-04T17:58:10.0420613Z  2.06935140e+32 2.15092710e+31]
2023-01-04T17:58:10.0420967Z dtype = float32, shape = (10, 9, 1)
2023-01-04T17:58:10.0421329Z Mismatched elements: 90 / 90 (100%)
2023-01-04T17:58:10.0421711Z Max absolute difference: 2.6462097e+36
2023-01-04T17:58:10.0422088Z Max relative difference: 1.
2023-01-04T17:58:10.0422591Z  x: array([[[ 6.941835e-01],
2023-01-04T17:58:10.0422963Z         [ 5.160517e-02],
2023-01-04T17:58:10.0423270Z         [ 1.027888e+00],...
2023-01-04T17:58:10.0423635Z  y: array([[[-2.354915e+36],
2023-01-04T17:58:10.0424006Z         [-3.053497e+35],
2023-01-04T17:58:10.0424365Z         [-2.027860e+36],...
```

**Review by penpornk — APPROVED** [review]
cc: @TensorFlow-MKL

File: tensorflow/core/kernels/mkl/mkl_einsum_op.cc [hunk 1]
```diff
@@ -115,14 +115,57 @@ struct MklEinsumHelper {
...
+      case 5:
+        weight_format =
+            trans_y ? memory::format_tag::abced : memory::format_tag::abcde;
+        break;
+      default:
+        weight_format = memory::format_tag::undef;
+    }
+    engine cpu_engine = engine(engine::kind::cpu, 0);
+    MklDnnData<T> weights_mkl(&cpu_engine);
+    if (weight_format != memory::format_tag::undef) {
```
Comment (mdfaijul) [inline tensorflow/core/kernels/mkl/mkl_einsum_op.cc:144]: What is the implication if `weight_format == memory::format_tag::undef`?

File: tensorflow/core/kernels/mkl/mkl_einsum_op.cc [hunk 2]
```diff
@@ -115,14 +115,57 @@ struct MklEinsumHelper {
...
+      auto weight_md =
+          memory::desc(params->b_dims, MklDnnType<T>(), weight_format);
+      std::shared_ptr<dnnl::matmul::primitive_desc> matmul_pd =
+          matmul_prim->GetPrimitiveDesc();
+      // Reorder weights if necessary.
+      // Check whether we need to do reorder.
+      if (weight_md != matmul_pd->weights_desc()) {
```
Comment (mdfaijul) [inline tensorflow/core/kernels/mkl/mkl_einsum_op.cc:151]: Does this condition gets true ever?

File: tensorflow/core/kernels/mkl/mkl_einsum_op.cc [hunk 3]
```diff
@@ -115,14 +115,57 @@ struct MklEinsumHelper {
...
             *params, false /* value for do_not_cache */);
 
+    T* weight_data = const_cast<T*>(rhs.flat<T>().data());
+
+#ifdef DNNL_AARCH64_USE_ACL
+    memory::format_tag weight_format;
+    switch (params->b_dims.size()) {
+      case 2:
+        weight_format =
+            trans_y ? memory::format_tag::ba : memory::format_tag::ab;
```
Comment (mdfaijul) [inline tensorflow/core/kernels/mkl/mkl_einsum_op.cc:125]: I think transpose situations are already handled by Line 110 for up to 12D cases, not only up to 5D.

**Pull request merged**